Move buildFileTree into a fileTree reducer

bobsilverberg opened the pull request:
Fixes #533 

This patch moves all of the functionality that creates the `tree` structure out of the `FileTree` component and into a new `fileTree` reducer. This includes:

- The types for `FileNode`, `DirectoryNode` and `TreeNode`.
- The functions `getVersionName`, `getRootPath`, `recursiveSortInPlace` and `buildFileTree`, all of which are used for generating the `tree` and were all moved exactly as is.
- The `FileTree` component now gets its `tree` prop via `mapStateToProps` from the `fileTree` reducer.
- The `FileTree` component now has a `loadData` function which is called from both `componentDidMount` and `componentDidUpdate` which checks whether has a value in the `tree` prop and if not dispatches `fileTreeActions.buildTree`.
- The new `fileTree` reducer exposes one action: `buildTree` which accepts a version and stores a `tree` structure in its state, and one selector: `getTree` which returns either the `tree` structure, or `undefined` if it hasn't been built yet for the requested `version`.

bobsilverberg commented:
My apologies for the large diff, but this is pretty much the minimum amount of changes required to move this function. The large number of lines changed has to do with the fact that quite a bit of code was moved from `FileTree` (component) to `fileTree` (reducer), and the same was done for the tests (a wholesale move of a lot of code from tests for the component to tests for the reducer). The diff for the test is super ugly though. :-(

willdurand commented on src/components/FileTree/index.spec.tsx:
```diff
@@ -1,399 +1,120 @@
     it('renders a ListGroup component with a Treefold', () => {
       const store = configureStore();
       const version = getVersion({ store });
+      _buildTree(store, version);
```
Could this be moved in `render()`? It looks like it is used everywhere.

bobsilverberg replied:
There are a couple of issues with that, which is why I didn't do it.

1. We don't always want to build the file tree, so I would have needed an argument like `buildFileTree: boolean`.
2. The `render` function is designed to accept a `versionId`, not a `version`. It's up to the test to decide whether to load a `version` that matches the `versionId` into the store before calling `render`. `buildTree` requires an actual `version`, so it would require changes to the `render` function, which is currently working pretty well for the tests we need.

I did experiment with this, but decided that keeping `render` the way it is would be best. Suggestions for improvements are welcome, though.

kumar303 commented on src/reducers/fileTree.tsx:
```diff
@@ -0,0 +1,179 @@
+            currentNode = foundNode;
+          }
+
+          // TODO: this should not happen but what if we don't find a node?
```
I know this was in the original patch but I think it would be nice if this threw an error or logged a warning. I'm imagining that hitting this line would result in a broken tree so it would be helpful to at least get some kind of signal of what went wrong. You could probably inject a stub `root` to test it -- whatever works.

kumar303 commented on src/reducers/fileTree.tsx:
```diff
@@ -0,0 +1,179 @@
+
+export const getTree = (
+  treeState: FileTreeState,
+  version: Version | void,
```
can this just be `versionId`?

kumar303 commented on src/reducers/fileTree.tsx:
```diff
@@ -0,0 +1,179 @@
+
+export const actions = {
+  buildTree: createAction('BUILD_TREE', (resolve) => {
+    return (payload: { version: Version }) => resolve(payload);
```
I think it's better to keep actions light whenever possible since they live on the stack of all user activity. What about accepting `versionId` and `addonName` (or `rootName`)?

bobsilverberg replied:
I'm not sure I follow. `buildTree` needs a version to work with to build the tree. Are you suggesting we get that from the `versions` reducer, rather than passing it in? Is it okay to call a selector from `versions` when inside the `fileTree` reducer?

kumar303 replied:
Where does it need a `version`? I only saw two places:

- to build the root name: https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L45
- to use `version.id` as a key https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L170

Thus, I think the action payload can just be `dispatch(buildTree({ versionId: 123, rootName: 'some-name' }))`.

Am I missing some other use of `version`?

willdurand replied:
Aren't we supposed to use the `entries` too? https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L53

bobsilverberg replied:
Yes, the entries are kind of the point of the whole thing. 😉

bobsilverberg commented:
Thanks for the initial reviews @kumar303 and @willdurand. I addressed the first two of @kumar303's comments, and added some replies to the other comments.

kumar303 commented:
> Aren't we supposed to use the entries too? 

Yes! That's what I was missing. OK, I guess you might as well pass `version` into the action. Carry on.
